### DISCUSSION TITLE
Utilities classes collisions

### DISCUSSION THREAD

**Discussion by edygar** [tailwindlabs/tailwindcss#5655] [Ideas]
Is there an utility that given some class such as `m-1 my-2 m-3 mb-3`, it could return only `m-3 mb-3`?

[There's this thing](https://twitter.com/mxstbr/status/1038416725182697472?lang=en) many people don't realize about CSS class names, their order in the list doesn't influence the final styling. That doesn't seem to be a problem, we can always spot this redundancy or conflicting classes by using a [linter](https://blog.tailwindcss.com/introducing-linting-for-tailwindcss-intellisense), but when we start composing class names dynamically, we may end up losing this control.

For example in React:
```jsx
const SomeComponent = ({ className }) => 
  <div className={`border-4 ${className}`}>
    {/* … lots of other components */}
  </div>

// Consumption
<SomeComponent className="border-2" />
```

In this case, SomeComponent ends up with `class="border-4 border-2"`, but applying solely `border-2`. The consumer of the component doesn't have consistency, because if it had defined className as `border-8`, it would end up with a thicker border. If we had a `tw` that could solve this redundancy prioritizing right to left, it could amplify the composability way much.

**Comment by edygar** [comment]
I've tried to write something of the sort, but it seems too heavy and I feel like it's the wrong approach. It should base on the `tailwind.config.js`, if anyone wants to take a look, here it is:

https://gist.github.com/edygar/6e95d40b99d00d41b0e4c268d1c2c5a1

It uses `classnames` to combine classes, just so it facilitates it's usage within React's context

**Comment by adamwathan** [comment]
Check this out by @dcastil, might be what you’re looking for!

https://github.com/dcastil/tailwind-merge
